### PR TITLE
Add ray grouped runtime

### PR DESCRIPTION
**Pull request by porteratzo**
Start from PR-791
created "ray_grouped" runtime backend
***openfl/experimental/runtime/local_runtime***
changed backend test to backend.find('ray') instead of backend=='ray' so that both backends work
In __get_collaborator_object add logic for if backend=="ray_grouped" where we use funtion ray_group_assign. Takes collaborators and number of collaborators per group and creates the RayGroup actors and returns GroupMember objects which manage the references of a collaborator to its respective group, allowing the original code in RayExecutor and LocalRuntime.execure_colabb_task to work without modification.

***openfl/experimental/utilities/runtime_utils.py***
changed the check_resource_allocation since I found logic errors that would make the funtion return failure anytime the sum of collaborator gpu_requirements is larger than 1

### CODE REVIEW

File: openfl/experimental/runtime/local_runtime.py [hunk 1]
```diff
@@ -62,6 +66,124 @@ def ray_call_get(self) -> List[Any]:
         return clones
 
 
+def ray_group_assign(collaborators, collaborators_per_group=3, max_concurrency=1):
+    # This function takes the list of collaborators and the collaborators per group and assigns the collaborators to resource groups which share cuda context,
+    # it returns a list of GroupMembers, that is a utility class that manages the collaborator and in which group it is in. This maintians compatablility with current runtime execution without having to manage the groups manually.
+    
+    class GroupMember:
+        # Utility class that maintains compatability with runtime execution,
```
Comment (psfoley) [inline openfl/experimental/runtime/local_runtime.py]: It's great you've added these comments. I would recommend converting these to docstrings so they're available in our upcoming revamped API documentation.

File: openfl/experimental/runtime/local_runtime.py [hunk 2]
```diff
@@ -62,6 +66,124 @@ def ray_call_get(self) -> List[Any]:
...
+def ray_group_assign(collaborators, collaborators_per_group=3, max_concurrency=1):
+    # This function takes the list of collaborators and the collaborators per group and assigns the collaborators to resource groups which share cuda context,
+    # it returns a list of GroupMembers, that is a utility class that manages the collaborator and in which group it is in. This maintians compatablility with current runtime execution without having to manage the groups manually.
+    
+    class GroupMember:
+        # Utility class that maintains compatability with runtime execution,
+        # It takes the collaborator_actor, and the name of a collaborator. it takes care of managing which group the collaborator is a member of. 
+        # It also assigns attributes for each funtion in the Collaborator interface in conjuntion with RemoteHelper to maintain compatability with RayExecutor.
+        def __init__(self, collaborator_actor, collaborator):
+            interface_module = importlib.import_module(
```
Comment (psfoley) [inline openfl/experimental/runtime/local_runtime.py]: Is it necessary to use `importlib` here? Could this be simplified to:

```
from openfl.experimental.interface import Collaborator
all_methods = [
                method
                for method in dir(Collaborator)
                if callable(getattr(Collaborator, method))
            ]
```

File: openfl/experimental/runtime/local_runtime.py [hunk 3]
```diff
@@ -1,232 +1,613 @@
...
+    def __init__(self):
+        self.collaborators = {}
+
+    def append(
+        self,
+        name: str = "",
+        private_attributes_callable: Callable = None,
+        **kwargs,
+    ):
+        interface_module = importlib.import_module("openfl.experimental.interface")
```
Comment (psfoley) [inline openfl/experimental/runtime/local_runtime.py]: This can again be simplified to:

`from openfl.experimental.interface import Collaborator`

Comment (porteratzo) [reply openfl/experimental/runtime/local_runtime.py]: Changed to from openfl.experimental.interface import Collaborator

File: openfl/experimental/runtime/local_runtime.py (hunk 2 above)
Comment (porteratzo) [reply openfl/experimental/runtime/local_runtime.py]: Changed to from openfl.experimental.interface import Collaborator

File: openfl/experimental/runtime/local_runtime.py (hunk 1 above)
Comment (porteratzo) [reply openfl/experimental/runtime/local_runtime.py]: Added docstrings

File: openfl/experimental/runtime/local_runtime.py [hunk 4]
```diff
@@ -97,22 +219,27 @@ def some_collaborator_task(self):
                                              collaborator tasks will run sequentially.
         """
         super().__init__()
-        if backend not in ["ray", "single_process"]:
+        if backend not in ["ray", "single_process", "ray_grouped"]:
```
Comment (psfoley) [inline openfl/experimental/runtime/local_runtime.py]: The changes you've added should become the `ray` backend. The inclusion of two backends that use `ray` will confuse users. Relabeling `ray_grouped` to `ray` will also automatically trigger workflow API tests to give more confidence on it's implementation

File: openfl-tutorials/experimental/Workflow_interace_201.py
Comment (psfoley) [inline openfl-tutorials/experimental/Workflow_interace_201.py:1]: This file should be removed - and any necessary changed should be made to `Workflow_Interface_201_Exclusive_GPUs_with_Ray.ipynb` (I think they already are?)

File: openfl/experimental/runtime/local_runtime.py (hunk 4 above)
Comment (porteratzo) [reply openfl/experimental/runtime/local_runtime.py]: replaced ray with raygrouped logic

File: openfl-tutorials/experimental/Workflow_interace_201.py
Comment (porteratzo) [reply openfl-tutorials/experimental/Workflow_interace_201.py:1]: removed unnecessary files

**Review by psfoley — APPROVED** [review]
Great work on this, @porteratzo (and @ParthM-GitHub for initial PR this subsumes). All of the recommended changes have been made. Approved.

**Pull request merged**